### DISCUSSION TITLE
The dreadful error "failed because the page updated" - shouldn't it do not exist in v12?

### DISCUSSION THREAD

**Discussion by dziedrius** [cypress-io/cypress#27490] [Questions and Help]
There was an issue with whooping number of comments - https://github.com/cypress-io/cypress/issues/7306, which states:
> Before these changes, Cypress stored a single DOM element (or window) reference that is passed between a set of commands to assert on, which could become stale as the App under test updated.
>
> With these changes, Cypress will now re-query for the DOM element (window object) to ensure its always the most up-to-date reference to verify

Given this, I was expecting that the "failed because the page updated" should be gone forever, seems it is not the case.
In current documentation, there's reference about the error: 
https://docs.cypress.io/guides/references/error-messages#cy-failed-because-the-page-updated
which is really nice, because it explains in quite detail what is going on and even provides an example.

What I did not like, was this:
>Queries (.get(), .as() and.parent(), for example) and assertions (.should(), .and()) are safe to chain off of. Commands (such as .click()) are not.

1. If something is not safe, wouldn't it be nice to provide full list, because right now I need... to guess? try? but this is a race condition, how much trying is enough?
2. Shouldn't API terminate chaining on unsafe command?

In my case, it was failing on seemingly simple chain:
> cy.get('[data-test="row"]').should('contain', 'test string').eq(0).click();

So I thought, may be `eq` is not a safe command, may be I should do a scientific experiment to prove it.
So I took example from help:
```html
<!DOCTYPE html>
<html>

<head>
    <script src="https://code.jquery.com/jquery-3.7.0.js"></script>
</head>

<body>
    <div data-testid="parent">
        <button><span>Delete</span></button>
    </div>
    <script>
        $('button').click(function () {
            // when the <button> is clicked
            // we remove the button from the DOM,
            // and add a new, identical one.
            $(this).replaceWith(this.outerHTML)
        })
    </script>
</body>

</html>
```
And put it to the test from help too, first as it was originally in help:
```js
describe('template spec', () => {
  it('passes', () => {
    cy.visit('http://localhost:3300/index.html');
    cy.get('button').click().parent();
  })
})
```
And it failed, just as described in documentation.   
Then tried to mimic my situation and replaced line with click with my case:
```
    cy.get('button').should('contain', 'Delete').eq(0).click();
```
And it passes, so it seems that `eq` is safe command after all. But it does not pass in my actual test suite, even though it was passing with cypress v11.

P.S. I am sorry if this looks like a rant (it actually is a rant after first week with cypress in my life, I need to steam off), hopefully this is a good source for discussion too.

**Comment by portyu9** [comment]
Hi @dziedrius 

Cypress 12's query/retry changes did not make every detached-element race impossible.

In your example:

```js
cy.get('[data-test="row"]')
  .should('contain', 'test string')
  .eq(0)
  .click()
```

`.eq()` itself is a retryable query and is not inherently unsafe.

The important part is that an assertion in the middle of a chain can become a retry boundary. Once that assertion succeeds, the subject can effectively be locked in. If your framework then rerenders and replaces that DOM node, subsequent work may be operating on the old detached element.

A safer pattern for a rerendering UI is to split the chain and re-query:

```js
cy.get('[data-test="row"]').should('contain', 'test string')

cy.get('[data-test="row"]')
  .eq(0)
  .click()
```

Also synchronize on whatever actually indicates that rendering has finished—an intercepted network request, application state, or a meaningful DOM assertion—rather than using an arbitrary wait.

After an action that causes a rerender, re-query before doing further work instead of continuing to chain from the pre-action element.

`force: true` is not a fix for a detached node; it only bypasses actionability checks.

Please mark this answer as accepted if it helped, thank you.
